### PR TITLE
lnrpc: routing fee inaccuracy

### LINKED ISSUE #4515: Allow setting routing fees accurately using non-fractional values
### Background

In the `UpdateChannelPolicy` method, a `PolicyUpdateRequest` is formed to represent an update to a channel routing policy, or to the global policy set

The routing proportional fee is set using the `fee_rate` value, which is represented by a double

This can result in rounding errors, such as attempting to set a routing fee of 31337 by passing `fee_rate: 0.031337` results in a fee rate of 31336

Ideally the caller could pass a simple integer scalar for the fee rate in parts per million instead of a fractional value

### Your environment

* lnd v0.11.0-beta.rc2

### Steps to reproduce

1. Call the RPC method `UpdateChannelPolicy` and specify a fee rate 0.031337
2. Look up the channel in the graph using the `GetChanInfo` API
3. Compare the policy fee rate to the specified fee rate, there is a mismatch

### Expected behavior

The specified fee rate should be used by LND

### Actual behavior

An incorrect fee rate is used

### PR DESCRIPTION
**Pull request by ErikEk**
When updating the channel routing policy, we encounter an inaccurate precision error when calculating the routing fee. The issue stems from the way the IEEE 754 standard works. 
The solution here is to add a uint64 parameter (as mentioned in the issue) and keep the float64 fee_rate parameter but rounding the product of the base and fee rate.

Fixes #4515 and #6189

### CODE REVIEW

**Review by bhandras — COMMENTED** [review]
Changes LGTM, but a more complete fix could perhaps be adding a new `FeeRatePpm` member to the proto where we can set the free rate as an integer. WDYT? @alexbosworth

**Comment by alexbosworth** [comment]
> Changes LGTM, but a more complete fix could perhaps be adding a new `FeeRatePpm` member to the proto where we can set the free rate as an integer. WDYT? @alexbosworth

Yes setting rate as integer instead of decimal would be good

**Comment by ErikEk** [comment]
I'll look into it.

**Review by carlaKC — COMMENTED** [review]
+1 for setting fee rate with an integer

**Comment by bhandras** [comment]
@ErikEk would you like to finish this PR? If you don't have time we can take it over (see also related bug: https://github.com/lightningnetwork/lnd/issues/6189)

**Comment by ErikEk** [comment]
@bhandras Let me take a look at it tonight. I've not been convinced we needed this until now. I'll submit a pr before Sunday. Will msg you tomorrow.

**Comment by bhandras** [comment]
> @bhandras Let me take a look at it tonight. I've not been convinced we needed this until now. I'll submit a pr before Sunday. Will msg you tomorrow.

Awesome, thanks @ErikEk :)

**Comment by Roasbeef** [comment]
Agree would be nice to get this into 0.14.2. I think we also want some _itests_ to validate the external consistency here w.r.t what we input and what we'll actually send out on the network.

**Comment by ErikEk** [comment]
@bhandras Updated the code as you wished. I'm working from abroad with limited resources right now, and not accessible everyday. If you want to take over the pr (because of the 0.14.2 release) I would take no offence. I haven't added any itests yet.

**Comment by bhandras** [comment]
> Updated:
My thinking here is to keep fee_rate as it used to work for backward compatibility but add fee_rate_ppm as an alternative interface option.

I see your point @ErikEk. I still think that backward compatibility is not required as it's a miscalculation for the floating point case.

**Comment by ErikEk** [comment]
@bhandras Just so I understand you correct - you don't mean we should replace the float fee_rate parameter with integer parameter fee_rate_ppm?

**Comment by bhandras** [comment]
> @bhandras Just so I understand you correct - you don't mean we should replace the float fee_rate parameter with integer parameter fee_rate_ppm?

I mean the float calculation should be corrected as in your original version. The newly added ppm param is a very nice addition!

**Comment by ErikEk** [comment]
@bhandras Updated the pr. The float parameter fee rate now gives the correct results.

File: lnrpc/lightning.proto [hunk 1]
```diff
@@ -3747,9 +3747,14 @@ message PolicyUpdateRequest {
     // The base fee charged regardless of the number of milli-satoshis sent.
     int64 base_fee_msat = 3;
 
-    // The effective fee rate in milli-satoshis. The precision of this value
-    // goes up to 6 decimal places, so 1e-6.
-    double fee_rate = 4;
+    oneof fee {
+        // The effective fee rate in milli-satoshis. The precision of this value
+        // goes up to 6 decimal places, so 1e-6.
+        double fee_rate = 4;
```
Comment (Roasbeef) [inline lnrpc/lightning.proto:3753]: This breaks the API as is. Instead I think we should just add this as a new field, then switch to the `oneof` later (particularly as we're planning on including this in a minor release).

**Review by Roasbeef — CHANGES_REQUESTED** [review]

File: lnrpc/lightning.proto (hunk 1 above)
Comment (bhandras) [reply lnrpc/lightning.proto:3753]: Maybe I'm missing something but moving an existing field into a `oneof` should not break backwards compatibility (see https://developers.google.com/protocol-buffers/docs/proto3#tag_reuse_issues)

**Comment by bhandras** [comment]
I'm working on some itests in the mean time so we can cherry pick together the final PR if @ErikEk would be unable to continue before 0.14.2 deadline.

**Comment by Roasbeef** [comment]
Replaced by #6198

**Pull request closed without merging**